homeassistant/components/reolink/host.py

```diff
@@ -379,10 +430,45 @@ def unregister_webhook(self):
         webhook.async_unregister(self._hass, self.webhook_id)
         self.webhook_id = None
 
+    async def _async_long_polling(self, *_) -> None:
+        """Use ONVIF long polling to immediately receive events."""
+        # This task will be cancelled once _async_stop_long_polling is called
+        while True:
+            if self._webhook_reachable:
+                self._long_poll_task = None
+                await self._async_stop_long_polling()
+                return
+
+            try:
+                channels = await self._api.pull_point_request()
+            except Exception as ex:  # pylint: disable=broad-except
+                if not self._long_poll_error:
+                    _LOGGER.exception("Error while requesting ONVIF pull point: %s", ex)
+                    await self._api.unsubscribe(sub_type=SubType.long_poll)
+                self._long_poll_error = True
+                await asyncio.sleep(30)
+                continue
+
+            self._long_poll_error = False
+
+            if not self._long_poll_received and channels != []:
+                self._long_poll_received = True
+                ir.async_delete_issue(self._hass, DOMAIN, "webhook_url")
+
+            # After receiving the new motion states in the upstream lib,
+            # update the binary sensors with async_write_ha_state
+            # The same dispatch as for the webhook can be used
+            if channels is None:
+                async_dispatcher_send(self._hass, f"{self.webhook_id}_all", {})
+                continue
+
+            for channel in channels:
+                async_dispatcher_send(self._hass, f"{self.webhook_id}_{channel}", {})
```

Comment:
I seem to recall this code exists in other places. Can this be made more DRY?

Reply:
What do you mean with "DRY"?

Yes it exists 3 times, will split it out to a seperate function.

Reply:
I have made a new function `_signal_write_ha_state` and that is now used 3 times.

Reply:
> What do you mean with "DRY"?

https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

But I think you figured that out 👍